roachtest: use toxiproxy for network testing

tbg opened the pull request:
Reincarnation of https://github.com/cockroachdb/cockroach/pull/23141.

[toxiproxy] is a TCP proxy which exports an HTTP API that allows
introducing various anomalous network conditions for resilience testing.

Among the interesting things one can simulate are latency and bandwidth
restrictions. This is done programmatically and thus lends itself to the
kind of testing we'd like to do; in particular, it should allow testing
locally a good approximation of geo-replicated clusters and various
network partitions.

The test introduced here isn't really useful yet. We'll need to settle
on something straightforward to test.

Updates #21680.
Updates #14768.

[toxiproxy]: https://github.com/Shopify/toxiproxy

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/33282)
<!-- Reviewable:end -->

tbg commented:
@petermattis finally got around to brushing this up. I still want to make this test useful before I merge it and expose infra that's a little less ad-hoc.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33282)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/network.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LU71kcLEOzZ3vdZyjFT:-LU71kcLEOzZ3vdZyjFU:bt3ztq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/028bbcd0c50b9949b41c546741f03f75aeb15682/pkg/cmd/roachtest/network.go#L69)):*
> ```Go
> echo "original args: $@"
> echo "modified args: ${args[@]}"
> ./cockroach.real "${args[@]}"
> ```

This script is unfortunate. I can't think of a better way to do this, though. Perhaps `roachprod` could be extended to make this easier. 
___
*[pkg/cmd/roachtest/network.go, line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LU7-iW5B1dKmQbfJcIz:-LU7-iW5B1dKmQbfJcJ-:b-vxkhem) ([raw file](https://github.com/cockroachdb/cockroach/blob/028bbcd0c50b9949b41c546741f03f75aeb15682/pkg/cmd/roachtest/network.go#L103)):*
> ```Go
> 	cockroachWrapper, undo1 := stringToTemp(t, cockroachToxiWrapper, 0755)
> 	defer undo1()
> 	toxiWrapper, undo2 := stringToTemp(t, toxiServerWrapper, 0755)
> ```

Rather than this `stringToTemp` approach, I think we should add a `cluster.PutFile` method (though that name needs improvement) that internally writes the file locally and then uploads it. We've need similar functionality in other tests. 
___
*[pkg/cmd/roachtest/network.go, line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LU70LbB04kYJk2M-Z8G:-LU70LbB04kYJk2M-Z8H:bt8ismf) ([raw file](https://github.com/cockroachdb/cockroach/blob/028bbcd0c50b9949b41c546741f03f75aeb15682/pkg/cmd/roachtest/network.go#L106)):*
> ```Go
> 	defer undo2()
> 
> 	c.Put(ctx, cockroachWrapper, "./cockroach", c.All())
> ```

I wonder if it would be clearer to leave the cockroach binary named `cockroach`. You can pass `startArgs("--binary=foo")` to `cluster.Start()` to have it run an arbitrary binary instead of `cockroach`.
___
*[pkg/cmd/roachtest/network.go, line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LU70f_4-qtWC02wHQ8W:-LU70f_4-qtWC02wHQ8X:b-4jqb52) ([raw file](https://github.com/cockroachdb/cockroach/blob/028bbcd0c50b9949b41c546741f03f75aeb15682/pkg/cmd/roachtest/network.go#L173)):*
> ```Go
> 				return err
> 			}
> 		}
> ```

This looks a lot like `waitForFullReplication`. Not quite the same. I don't have a concrete suggestion for unifying them. Do you mind explaining what the differences are for?


<!-- Sent from Reviewable.io -->

tbg commented:
I have a first question during testing this. I'm running a four node cluster. All nodes have 150ms upstream and downstream latencies added via toxiproxy, except for the first node. Now look at this:

![image](https://user-images.githubusercontent.com/5076964/50284127-50c17680-0458-11e9-8d48-bdb7ef7c2be2.png)

Everything but the first column makes sense to me. For example, to send a message from n2 to n3 and back, you eat 150ms and another 150s. But how do you explain the ~250ms totals when n1 is involved? I would have expected ~150ms.

tbg commented:
Actually, that's not quite right. For the first row, I think it means n1 contacted n2, so the RPC eats the incoming and outgoing delay on n2 for a total of 300ms. So the first row makes sense.

First column should be different, though. This time, nX is the sender and n1 the receiver, so there should't be any delay at all.

petermattis reviewed (approved):
Nice. Did you figure out why you were seeing 250ms latencies?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-:-LUBAN0G6K2tTsuCMt_i:brw53w3)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg commented:
No, it also hasn't happened since. I was playing around with toxiproxi-cli on that node though, so.. not sure. Either way I'll put a useful test in here that tests some stuff under network partitions before I merge. I'll ping you.

tbg commented:
Under a nemesis that lets incoming connections to n1 time out and has everything else work, we're definitely leaking goroutines (`master`) on that cut-off node:

```
$ cat ~/local/1/logs/cockroachreal.log | grep ' goroutines'
I181220 14:17:27.331042 341 server/status/runtime.go:464  [n1] runtime stats: 126 MiB RSS, 255 goroutines, 189 MiB/58 MiB/269 MiB GO alloc/idle/total, 19 MiB/26 MiB CGO alloc/total, 0.0 CGO/sec, 0.0/0.0 %(u/s)time, 0.0 %gc (9x), 6.0 MiB/6.0 MiB (r/w)net
...
I181220 16:46:53.958230 341 server/status/runtime.go:464  [n1] runtime stats: 112 MiB RSS, 526 goroutines, 168 MiB/12 MiB/220 MiB GO alloc/idle/total, 13 MiB/26 MiB CGO alloc/total, 14.4 CGO/sec, 1.4/1.2 %(u/s)time, 0.0 %gc (0x), 436 KiB/436 KiB (r/w)net
```

and the rate of leaks is faster on the other nodes:

```
I181220 16:48:25.603964 216 server/status/runtime.go:464  [n4] runtime stats: 105 MiB RSS, 1687 goroutines, 95 MiB/25 MiB/166 MiB GO alloc/idle/total, 13 MiB/24 MiB CGO alloc/total, 10.3 CGO/sec, 1.0/0.8 %(u/s)time, 0.0 %gc (0x), 407 KiB/408 KiB (r/w)net
```

tbg commented:
Besides, the UI and SQL are completely hosed. I mean completely, basically nothing works. I assume this is because n1 can heartbeat its liveness record just fine, but it can't actually be contacted for anything.

petermattis commented:
Progress!

tbg commented:
(This "nothing works" includes the node diagnostics page, whose only job is basically... showing the pings between the nodes)

tbg commented:
Heh, yeah. n1 is leading the liveness range just fine.

![image](https://user-images.githubusercontent.com/5076964/50298962-b0cc1300-0480-11e9-98ab-3853f77ebd8d.png)

If any other node tried to propose anything, that'd end up on the raft stream in the other direction, which is blocked. And consequently we're seeing pending proposals on these nodes, which I checked.

tbg commented:
Here's the goroutine leak on the non-partitioned nodes:

```
github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal

/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go

  Total:           0       9514 (flat, cum) 97.50%
    350            .          .           		ctxDone = nil 
    351            .          .           	} 
    352            .          .            
    353            .          .           	// Wait for the inflight request. 
    354            .          .           	var res singleflight.Result 
    355            .       9514           	select { 
    356            .          .           	case res = <-resC: 
    357            .          .           	case <-ctxDone: 
    358            .          .           		return nil, nil, errors.Wrap(ctx.Err(), "aborted during range descriptor lookup") 
    359            .          .           	} 
    360            .          .            
```

On the partitioned node, we have instead

![image](https://user-images.githubusercontent.com/5076964/50334830-c9344000-0509-11e9-94ec-a5574bf22c4d.png)

tbg commented:
What's interesting here is that we're seeing `rollback` in the trace, similar to how we saw `tryAsyncAbort` in a privately reported issue of the same nature.

tbg committed:
client: set timeout on async rollback's context

This fixes a goroutine leak under long-running network partitions.

Release note (bug fix): Fix a goroutine leak that would occur while
a cluster was unavailable (or a subset of nodes partitioned away
from the cluster) and would cause a resource spike to resolve.

tbg commented:
Well, who knew? Resolving all the toxics, I noticed my laptop's fan going crazy, and lo and behold the goroutine count had skyrocketed:

/Users/tschottdorf/local/1/logs/cockroachreal.toblerone.tschottdorf.2018-12-20T14_17_17Z.007173.log:I181221 09:52:58.661508 341 server/status/runtime.go:464  [n1] runtime stats: 623 MiB RSS, 4566 goroutines, 262 MiB/157 MiB/603 MiB GO alloc/idle/total, 47 MiB/71 MiB CGO alloc/total, 6076.0 CGO/sec, 43.8/6.9 %(u/s)time, 0.0 %gc (3x), 58 MiB/58 MiB (r/w)net
/Users/tschottdorf/local/1/logs/cockroachreal.toblerone.tschottdorf.2018-12-20T14_17_17Z.007173.log:I181221 09:53:08.902404 341 server/status/runtime.go:464  [n1] runtime stats: 1.2 GiB RSS, 28182 goroutines, 743 MiB/71 MiB/1.4 GiB GO alloc/idle/total, 47 MiB/70 MiB CGO alloc/total, 1919.1 CGO/sec, 450.8/76.2 %(u/s)time, 0.0 %gc (2x), 38 MiB/38 MiB (r/w)net

tbg commented:
The count came back down after a short moment, but this was reminiscent of what we saw in that privately reported issue as well. Likely the amount of work that had gathered up caused a further degradation. Unfortunately, by the time I had grabbed a goroutine profile, it was back to normal on all nodes, but this should be reproducible. I'll note that if we had something similar to #26797 then we'd have the info for free.

petermattis commented:
@tbg Looks like you have a solid grasp on this issue. Let me know if you need an additional pair of eyes. In the meantime, I'll grab some 🍿 and sit back and watch.

tbg commented:
Updated, should turn green now. @petermattis could you review this today? A vendor bump is included.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I didn't grok all of the logic in the tests, but I did take a close look at the `ToxiCluster` implementation and API and liked it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-:-LWJboWeDIqP6Pd8ktg_:bagin8z)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/network.go, line 80 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWJJfjh9bPyI3Fzpm1g:-LWJJfjh9bPyI3Fzpm1h:b-c70agj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/network.go#L80)):*
> ```Go
> 		}
> 
> 		c.Measure(ctx, 1, "set tracing=on; insert into test.commit values(3,1000), (1,1000), (2,1000); select age, message from [ show trace for session ]")
> ```

Nit: line-wrap that long SQL statement and capitalize the keywords.

What is the value of enabling tracing for this test.
___
*[pkg/cmd/roachtest/network.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWINlQH2JXVJ_Q7LDnw:-LWINlQH2JXVJ_Q7LDnx:bc774xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/network.go#L243)):*
> ```Go
> 			runNetworkTPCC(ctx, t, c, numNodes)
> 		},
> 	})
> ```

There is a small advantage to adding a parent test with the cluster definition so that these two tests can share the cluster creation overhead. See the `acceptance` tests if you want to go that route. 
___
*[pkg/cmd/roachtest/toxiproxy.go, line 92 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWJcBW63L5RXml_oYaG:-LWJcBW63L5RXml_oYaH:b8pv83p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/toxiproxy.go#L92)):*
> ```Go
> 	// NB: we can use upstream once they've released 2.1.4; there's a pretty
> 	// dramatic perf problem with the "timeout" toxic in 2.1.3.
> 	toxiURL := "https://github.com/tbg/toxiproxy/releases/download/v2.1.4rc1/toxiproxy-server-linux-amd64"
> ```

Should we vendor this under `github.com/cockroachdb/` instead of `github.com/tbg`?
___
*[pkg/cmd/roachtest/toxiproxy.go, line 197 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWJceIgC8YBZcriD_eH:-LWJceIhE4NBQEam-8ZC:b-ozd0l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/toxiproxy.go#L197)):*
> ```Go
> // client connection) and measures the duration (including the invocation time
> // of `./cockroach sql`. This is simplistic and does not perform proper
> // escaping. It's not useful for anything but simple sanity checks.
> ```

I'm curious why you want through this hassle, rather than measuring the time for executing the SQL query in the roachtest itself. Is there a big overhead to doing the `ssh` into the remote node in order to run the query? I could also imagine opening a connection to the desired node and then measuring the time to run the query, which should involve no `ssh` overhead.


<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: use toxiproxy for network testing

Reincarnation of https://github.com/cockroachdb/cockroach/pull/23141.

[toxiproxy] is a TCP proxy which exports an HTTP API that allows
introducing various anomalous network conditions for resilience testing.

Among the interesting things one can simulate are latency and bandwidth
restrictions. This is done programmatically and thus lends itself to the
kind of testing we'd like to do; in particular, it should allow testing
locally a good approximation of geo-replicated clusters and various
network partitions.

The test introduced here isn't really useful yet. We'll need to settle
on something straightforward to test.

Updates #21680.
Updates #14768.

[toxiproxy]: https://github.com/Shopify/toxiproxy

Release note: None

tbg committed:
vendor: pick up toxiproxy client

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33282)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/network.go, line 80 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWJJfjh9bPyI3Fzpm1g:-LWPyJoc0H2eaymnqIiX:boq6lq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/network.go#L80)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: line-wrap that long SQL statement and capitalize the keywords.

What is the value of enabling tracing for this test.
</blockquote></details>

Done.

The test is basically a sanity check, and it has been helpful to see the internal timings.
___
*[pkg/cmd/roachtest/network.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWINlQH2JXVJ_Q7LDnw:-LWPyTrR6fH7ovN-DpW_:b-tgmo9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/network.go#L243)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There is a small advantage to adding a parent test with the cluster definition so that these two tests can share the cluster creation overhead. See the `acceptance` tests if you want to go that route. 
</blockquote></details>

I'm just going to skip the sanity check for now.
___
*[pkg/cmd/roachtest/toxiproxy.go, line 92 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWJcBW63L5RXml_oYaG:-LWQAWZ9E2DDQhRPP3eV:bgvjx49) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/toxiproxy.go#L92)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should we vendor this under `github.com/cockroachdb/` instead of `github.com/tbg`?
</blockquote></details>

Upstream's next release should include the fix, so I don't want to "officially" fork this at this point. Upstream actually has released 2.1.4, but without binaries: https://github.com/Shopify/toxiproxy/releases
Going to leave as is for now.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33282)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/toxiproxy.go, line 197 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/33282#-LWJceIgC8YBZcriD_eH:-LWQAqy3A_ivq8xzxJVD:bbxwxj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ee0d15917859927f609f761680c128ca540bc2/pkg/cmd/roachtest/toxiproxy.go#L197)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm curious why you want through this hassle, rather than measuring the time for executing the SQL query in the roachtest itself. Is there a big overhead to doing the `ssh` into the remote node in order to run the query? I could also imagine opening a connection to the desired node and then measuring the time to run the query, which should involve no `ssh` overhead.
</blockquote></details>

If you, as I do, have around a 100ms latency to the server, you'll want to run stuff from the server directly to get your measurements.
I agree that this is all a bit wonky., and it would be a lot nicer to be able to time a SQL query programmatically.


<!-- Sent from Reviewable.io -->

tbg commented:
TFTR!

bors r=petermattis

I'm sure this will fail tonight (though it just passed). Curious how it will manage to do it.

cc @andreimatei about the second commit.
